File: npm_api/src/results.ts
import { ProtocolError, ResultError } from "./errors.js";

/**
 * Returns a result union member when its __typename is one of success, and
 * throws ResultError for any other member (ValidationError, NotFoundError,
 * AuthError, RateLimitError, or a member this SDK does not know yet).
 *
 *     const stream = expectResult(data.createStream, "Stream");
 */
export function expectResult<T extends { __typename: string }, K extends T["__typename"]>(
  value: T | null | undefined,
  ...success: K[]
): Extract<T, { __typename: K }> {
  if (value === null || value === undefined) {
    throw new ProtocolError("result is missing");
  }
  if ((success as string[]).includes(value.__typename)) {
    return value as Extract<T, { __typename: K }>;
  }
  throw new ResultError(value as unknown as Record<string, unknown>);
}

File: sdk_python/src/livepeer_frameworks/results.py
"""Result unions: success members are returned, error members raised."""

from __future__ import annotations

from typing import Any, TypeVar, overload

from pydantic import BaseModel

from .errors import ProtocolError, ResultError

T = TypeVar("T", bound=BaseModel)


@overload
def expect_result(value: object, success: type[T], /) -> T: ...


@overload
def expect_result(value: Any, *success: str) -> Any: ...


def expect_result(value: Any, *success: Any) -> Any:
    """Returns a result union member when it is the success member, and
    raises ResultError for any other member (ValidationError, NotFoundError,
    AuthError, RateLimitError, or an UnknownMember, whose ResultError carries
    its __typename, message, and code). value is a generated model or raw response
    data; success is the generated success class, which narrows the type, or
    one or more __typename strings:

        stream = expect_result(created.create_stream, CreateStreamCreateStreamStream)
        stream = expect_result(created.create_stream, "Stream")
    """
    if value is None:
        raise ProtocolError("result is missing")
    typename = value.get("__typename") if isinstance(value, dict) else getattr(value, "typename__", None)
    for want in success:
        if (isinstance(want, type) and isinstance(value, want)) or want == typename:
            return value
    dumped: dict[str, Any] = value.model_dump(by_alias=True, mode="json") if isinstance(value, BaseModel) else dict(value)
    raise ResultError(dumped)
